### ISSUE TITLE
[Kimi K3] Add MXFP4-weight / MXFP8-activation QAT and checkpoint interoperability

### ISSUE THREAD

**Issue by elfiegg** [pytorch/torchtitan#4269]
## Parent

- Tracks #3029.
- Builds on #4025.

## What to build

Add an end-to-end Kimi K3 SFT/QAT path with:

- BF16 master parameters and optimizer state.
- Stateless 1x32 MXFP4 weight fake quantization.
- MXFP8 activations and grouped GEMMs for routed MoE experts.
- Import of the released packed MXFP4 Hugging Face checkpoint.
- Distributed export from a resumable DCP checkpoint to HF-compatible packed safetensors.

Resolve the quantized parameter set using the pinned HF [`quantization_config`](https://huggingface.co/moonshotai/Kimi-K3/blob/main/config.json#L431-L469), the HF module hierarchy, and the actual `weight_packed`/`weight_scale` pairs in the safetensors index. Do not hard-code an assumption that only `w1/w2/w3` are quantized.

Extend the existing Kimi K3 state-dict adapter while preserving its unquantized path:

```text
HF packed weights/scales
  -> dequantized per-expert weights
  -> TorchTitan grouped BF16 masters
  -> SFT/QAT
  -> split and repack
  -> HF packed weights/scales
```

MXFP8 activation quantization is a TorchTitan/TorchAO runtime behavior; it is not specified by the released checkpoint metadata.

### TorchAO ownership and training-time parameter behavior

TorchAO should own the model-independent numerical primitives: MXFP4 E2M1 fake quantization with 1x32 E8M0 scales and STE backward, plus the emulated/native MXFP8 grouped-GEMM path. TorchTitan should own model conversion, FQN selection, distributed composition, recipes, checkpoint loading, and export.

The converter swaps the routed-expert **execution path**, not the registered parameter object. During training, each expert weight remains a BF16 `nn.Parameter` updated normally by the optimizer. Every forward derives an MXFP4 quantize/dequantize view of that master, quantizes activations to MXFP8, and invokes grouped GEMM. STE backward routes the resulting gradient to the original BF16 master. Packed MXFP4 weights and E8M0 scales are materialized only for checkpoint import/export, never installed as optimizer parameters.

## Definition of success

- Kimi K3 debug and full-topology recipes can execute SFT with MXFP4-constrained expert weights and MXFP8 expert computation while updating BF16 master parameters.
- DCP checkpoints resume model, optimizer, scheduler, and training progress without storing packed weights or quantization scales as trainable state.
- The released packed checkpoint can initialize sharded BF16 masters without gathering all experts on one rank.
- Exported checkpoints contain exactly the expected packed/scaled and BF16 keys and reload through the official Kimi K3/compressed-tensors path.
- The implementation composes with FSDP and EP, including empty/ragged expert routing.
- Debug-model QAT remains numerically stable and does not materially regress convergence or compiled MXFP8 performance.

## Execution and validation milestones

1. **Freeze the checkpoint format**

   Pin the HF revision, enumerate every packed/scale pair, expand the target/ignore policy, and establish byte-compatible MXFP4 packing against `compressed-tensors`.

   **Gate:** no orphan keys or policy/manifest disagreements; packed codes and E8M0 scales match the reference serializer.

2. **Implement the training primitive**

   Add stateless MXFP4 fake quantization with STE backward and compose it with the existing MXFP8 grouped-GEMM path.

   **Gate:** codec tests are bit-exact; gradients reach only BF16 masters; emulated and native grouped-GEMM paths meet defined output and gradient tolerances.

3. **Integrate Kimi K3 QAT**

   Add a generic TorchTitan converter and Kimi debug SFT recipe without embedding quantization logic in the Kimi expert implementation. Preserve existing MXFP8-only and BF16 recipes.

   **Gate:** a seeded 200-step debug run has finite loss and gradients, decreasing tail loss, and no unexpected parameters or scales in the DCP state.

4. **Add packed checkpoint import and export**

   Extend the existing adapter to dequantize official packed expert tensors into grouped BF16 masters and reverse the mapping during distributed export.

   **Gate:** tiny official-style fixtures round-trip bit-exactly; a real released shard decodes correctly; exported safetensors reload with no missing or unexpected keys.

5. **Validate distributed training and resume**

   Exercise FSDP, EP, zero-token experts, ragged routing, and mixed image/text ranks, followed by a full-topology cluster smoke test.

   **Gate:** no collective timeout or full-expert gather; uninterrupted and DCP-resumed runs agree within the declared tolerance; compiled QAT stays within the agreed performance budget.

## Non-goals (yet)

- Learned quantization scales.
- Direct optimizer updates to packed MXFP4 parameters.
- Persistent MXFP8 master weights.
- Implementing EP itself; quantization should compose with the independent EP work.
- Quantizing ignored attention, shared-expert, vision, projector, or LM-head parameters.

## Blocked by

- The Kimi K3 model and unquantized adapter in #4025.

**Comment by elfiegg** [comment]
cc @shuhuayu @tianyu-l @syed-ahmed

**Comment by tianyu-l** [comment]
@shuhuayu we can create a tracking list like https://github.com/pytorch/torchtitan/issues/1118

**Comment by shuhuayu** [comment]
> [@shuhuayu](https://github.com/shuhuayu) we can create a tracking list like [#1118](https://github.com/pytorch/torchtitan/issues/1118)

sounds good, made one here: https://github.com/pytorch/torchtitan/issues/4272
